src/never.rs

```diff
@@ -0,0 +1,91 @@
+use {Future, Poll};
+
+/// A future which is never resolved.
+///
+/// This future can be created with the `empty` function.
+#[derive(Copy, Clone)]
+pub struct Never {}
+
+/// Creates a future which never resolves, representing a computation that never
+/// finishes.
+///
+/// The returned future will never resolve with a success but is still
+/// susceptible to cancellation. That is, if a callback is scheduled on the
+/// returned future, it is only run once the future is dropped (canceled).
+pub fn never<T, E>() -> impl Future<Item = T, Error = E> + Send + 'static {
```

Comment:
Ideally, the `!` type would only be coerced to a "real" type when necessary. I added the type parameters to the `never` function so that it would automatically coerce without requiring a `map` or `map_err`, but without default type parameters we don't have a way to make the coercion optional. Perhaps `Never` could have `coerce`, `coerce_error`, and `coerce_item` methods to make things easier?

As for why I like `!`, most of my reasons come directly [from the RFC](https://github.com/rust-lang/rfcs/pull/1216). TLDR: it can be used to detect dead result-handling code, to safely extract the result or error from a `Future`, and because `!` clearly documents that `Never` is a divergent future (it never returns). I also appreciate that there aren't different `Never<T, E>` types-- they're all the same.

Additionally, `!`/`Never`/`Nothing` is a common part of `Future::success/failed/never` abstractions in other languages (such as [this use of the `Nothing` type in RxScala observables](https://github.com/ReactiveX/RxScala/blob/0.x/src/main/scala/rx/lang/scala/Observable.scala#L4961)). Matching behavior from other languages allows developers to pick up the library more quickly.

Of course, all of these conceptual niceties fall on their face if `Never` isn't useful. At the end of the day, the point of this change is to make code cleaner and easier to understand. If `Never` doesn't accomplish that goal, then it's not a change worth making.

From what I can tell, `Never` as it currently stands is at least as good as `Empty`. Nightly feature dependencies and lack of default type parameters make it less convenient than it could be in the future. `Never` provides room for growth, where as `Empty` will always require explicit associated types.